File: frontend/src/game/upgrades/stat_modifiers/bullet_size_1.ts
import { Upgrade, type UpgradeDef } from '../Upgrade'
import { UpgradeTargetID, UpgradeStatID, RarityID, UpgradeTypeID } from '../../data/ID'

export const BulletSize1Def: UpgradeDef = {
  id: "bullet_size_1",
  name: "Heavy Rounds",
  description: "+10% bullet size",
  rarity: RarityID.Uncommon,
  upgradeType: UpgradeTypeID.StatModifier,
  cost: 6,
  targetClass: UpgradeTargetID.Bullet,
  fieldInTargetClass: UpgradeStatID.Size,
  value: 0.1,
  isMultiplier: true,
  specificAttackType: "bullet",
  stackable: true,
  maxStacks: 3,
}

export class BulletSize1 extends Upgrade {}

File: backend/app/core/upgrades/stat_modifiers/bullet_size_1.py
from ..upgrade_implementation import UpgradeImplementation, UpgradeAppliers


class BulletSize1(UpgradeImplementation):
    id = 'bullet_size_1'
    name = 'Heavy Rounds'
    description = '+10% bullet size'
    rarity = 'uncommon'
    type = 'stat_modifier'
    cost = 6

    def can_apply(self, state: dict) -> bool:
        return True

    def apply(self, appliers: UpgradeAppliers) -> None:
        value = 0.1
        appliers.modifiers.add_multiplier('bullet', 'size', value)

    def validate(self, recorded_stats: dict) -> bool:
        return True
